src/lib.rs

```diff
@@ -132,6 +137,140 @@ impl Invariant {
+
+        let mut lower_tick = self
+            .ticks
+            .get(pool_key, lower_tick)
+            .cloned()
+            .unwrap_or_else(|_| Self::create_tick(self, pool_key, lower_tick).unwrap());
+
+        let mut upper_tick = self
+            .ticks
+            .get(pool_key, upper_tick)
+            .cloned()
+            .unwrap_or_else(|_| Self::create_tick(self, pool_key, upper_tick).unwrap());
+
+        let (position, x, y) = Position::create(
+            &mut pool,
+            pool_key,
+            &mut lower_tick,
+            &mut upper_tick,
+            current_timestamp,
+            liquidity_delta,
+            slippage_limit_lower,
+            slippage_limit_upper,
+            current_block_number,
+            pool_key.fee_tier.tick_spacing,
+        )?;
+
+        self.pools.update(&pool_key, &pool)?;
+
+        self.positions.add(&caller, &position);
+
+        self.ticks.update(pool_key, lower_tick.index, lower_tick)?;
+        self.ticks.update(pool_key, upper_tick.index, upper_tick)?;
+
+        self.transfer_tokens(&pool_key.token_x, None, &caller, &program, x.get()).await?;
+        self.transfer_tokens(&pool_key.token_y, None, &caller, &program, y.get()).await?;
+
+        Ok(position)
+    }
+
+    pub fn get_position(&self, owner_id: &ActorId, index: u32) -> Result<Position, InvariantError> {
+        self.positions.get(owner_id, index).cloned()
+    }
+    
+    pub fn get_tick(&self, key: PoolKey, index: i32) -> Result<Tick, InvariantError> {
+        self.ticks.get(key, index).cloned()
+    }
+
+    pub fn is_tick_initialized(&self, key: PoolKey, index: i32) -> bool {
+        self.tickmap.get(index, key.fee_tier.tick_spacing, key)
+    }
+    
+    async fn transfer_tokens(
+        &mut self,
+        token_address: &ActorId,
+        tx_id: Option<u64>,
+        from: &ActorId,
+        to: &ActorId,
+        amount_tokens: u128,
+    ) -> Result<(), InvariantError> {
+        let tx_id = tx_id.or(self.get_transaction_id().into());
+        let reply = msg::send_for_reply_as::<_, Result<FTEvent, FTError>>(
+            *token_address,
+            FTAction::Transfer {
+                tx_id,
+                from: *from,
+                to: *to,
+                amount: amount_tokens,
+            },
+            0,
+            0,
+        )
+        .map_err(|_| InvariantError::TransferError)?
+        .await
+        .map_err(|_| InvariantError::TransferError)?;
+    
+        match reply {
+            Ok(ft_event) => {
+                match ft_event {
+                    FTEvent::Transfer { from: _, to: _, amount: _ } => {
+                        return Ok(())
+                    },
+                    _ => return Err(InvariantError::TransferError)
+                }
+            }
+            Err(_ft_error) => {
+                return Err(InvariantError::TransferError)
+            }
+    
+        }
+    }
+
+    fn get_transaction_id(&mut self) -> u64 {
```

Comment:
This name `get_transaction_id` is a bit misleading. Typically, a getter doesn't modify anything, it just returns value. I would suggest using a name like: `generate_transaction_id`/`assign_transaction_id`